### PR TITLE
Bump org.apache.logging.log4j:log4j-core from 2.17.1 to 2.20.0

### PR DESCRIPTION
**Pull request by Xtansia**
### Description
- Bump org.apache.logging.log4j:log4j-core from 2.17.1 to 2.20.0
- Use log4j's PluginProcessor annotation processor for plugin discovery
- Remove usage of deprecated explicit plugin discovery
- Pass through Configuration in OpenSearchJsonLayout to work around https://issues.apache.org/jira/browse/LOG4J2-3562

### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by reta** [comment]
@Xtansia there is another one on the same subject, https://github.com/opensearch-project/OpenSearch/pull/7586

**Comment by Xtansia** [comment]
> @Xtansia there is another one on the same subject, #7586

@reta That's my bad for not double checking before PR'ing, I'd gone down a rabbit hole figuring out the log4j errors you noticed in https://github.com/opensearch-project/OpenSearch/pull/8209#issuecomment-1609507938, which turned out to be plugin discovery issues not config issues.

I see that PR's been stalled since May due to the test failures, the code changes in this PR solves them. 

How would you prefer to handle the duped PRs? 

cc @owaiskazi19

**Comment by owaiskazi19** [comment]
@Xtansia I went ahead and close the previous PR.

**Review by Poojita-Raj — APPROVED** [review]

File: server/build.gradle [hunk 1]
```diff
@@ -173,6 +174,10 @@ tasks.withType(JavaCompile).configureEach {
   options.compilerArgs -= '-Xlint:unchecked'
 }
 
+compileJava {
```
Comment (reta) [inline server/build.gradle:177]: Sorry @Xtansia, to fill the gap in my understanding, adding `annotationProcessor` would not result into adding`-processor <...>` compiler argument out of the box? (I would have expected that to be fair)

**Review by reta — APPROVED** [review]

File: server/build.gradle (hunk 1 above)
Comment (Xtansia) [reply server/build.gradle:177]: So `annotationProcessor` adds onto `-processorpath` as Gradle deviates from `javac` behaviour and doesn't allow searching compile classpath for processors. Which normally the processors would then be automatically discovered from via javac, however our extension of the java plugin automatically adds `-proc:none` unless we're explicitly specifying processors https://github.com/opensearch-project/OpenSearch/blob/74788c3f45895a90d030b2434539beac7a1b1723/buildSrc/src/main/java/org/opensearch/gradle/OpenSearchJavaPlugin.java#L167-L169

Comment (Xtansia) [reply server/build.gradle:177]: You can see benchmarks needing to do the same thing here: https://github.com/opensearch-project/OpenSearch/blob/74788c3f45895a90d030b2434539beac7a1b1723/benchmarks/build.gradle#L49-L57

Comment (reta) [reply server/build.gradle:177]: Got it, thank you

**Comment by reta** [comment]
Hm ... interesting observation @Xtansia , we seems to generate less logs, on every Jenkins Gradle check we used to generate ~3.2Mb of logs

> Skipping 3,227 KB.. [Full Log](https://build.ci.opensearch.org/job/gradle-check/18616/consoleFull) 
> Skipping 3,267 KB.. [Full Log](https://build.ci.opensearch.org/job/gradle-check/18631/consoleFull)
> Skipping 3,323 KB.. [Full Log](https://build.ci.opensearch.org/job/gradle-check/18633/consoleFull)

But we are down to 2.5Mb with this change:

 > Skipping 2,586 KB.. [Full Log](https://build.ci.opensearch.org/job/gradle-check/18625/consoleFull)
 > Skipping 2,943 KB.. [Full Log](https://build.ci.opensearch.org/job/gradle-check/18612/consoleFull)

(nothing critical, just observation)

**Comment by reta** [comment]
@dblock @andrross  good to go?

**Review by dblock — APPROVED** [review]

**Comment by Xtansia** [comment]
@dblock @reta I think this is good to merge now?

**Comment by reta** [comment]
@Xtansia needs manual backport sadly

**Comment by MaxKsyunz** [comment]
Are planning to include this in 2.9?

**Pull request merged**